### PR TITLE
New pkgdown site, updated documentation, check env placeholders

### PR DESCRIPTION
**Pull request by gadenbuie**
This mostly adds a new pkgdown site with re-written documentation. But along the way a few non-documentation related changes sneaked in, described below.

You can preview the documentation at <https://friendly-golick-0771f8.netlify.app/>. Before merging this PR, I'll reset the pkgdown workflow so that we continue to publish to `gh-pages`. I'll separately work out the move to the rstudio org and publishing to Netlify.

I've also temporarily "removed" the `gradethis_demo()` by marking it as internal. I've written a _Gettting Started_ vignette that covers the same or similar material, so there's less of an immediate need.

#### Using sentinel placeholders for default values

Rather than using `missing_arg()` as a placeholder for `x`, `y`, or other places where a default value is automatically found in the checking environment, I've added sentinel objects, such as `.result` and `.solution`.

These objects appear as the default value in the function signature, which helpfully indicates what value the argument will take if a custom argument isn't provided.

I originally didn't export these, but exporting them gives us a consistent place to document their meaning. I also added a custom print method the prints out the definition of the object:

```r
.result
#> A placeholder for `.result` for use in `grade_this()`. The last value
#> returned from evaluating the user's exercise submission.
.solution_code
#> A placeholder for `.solution_code` for use in `grade_this()`. A string
#> containing the code provided within the `*-solution` chunk for the
#> exercise.
.envir_prep
#> A placeholder for `.envir_prep` for use in `grade_this()`. A copy of
#> the R environment after running the exercise setup code and before the
#> execution of the student's submitted code.
```

Generally, these placeholders are used by functions that are intended for use inside `grade_this()`. That means that if the _user_ writes `.result` in `grade_this({fail_if_equal(x = .result)})`, the object found will be `check_env$.result`. The placeholders mostly exist so we can signal our intention to find those objects in functions _within_ gradethis. I've updated all of the `graded()` helpers to use this pattern, as well as a few other places where `get0()` was used in a function signature, like `code_feedback()`.

#### Argument order in fail_if_code_feedback()

This is a minor change that makes `message` the first argument of `fail_if_code_feedback()`. It's unlikely that either `user_code` or `solution_code` would be customized, so they did not need to be the first two arguments.

### CODE REVIEW

File: R/grade_this.R [hunk 1]
```diff
@@ -158,41 +316,44 @@ detect_grade_this <- function(env = parent.frame()) {
...
-#' grade_this({
-#'   pass_if_equal(42, "Good stuff!")
-#' 
-#'   fail()
-#' })(submission)
-#' 
-#' # default fail() will show debug until you reset gradethis.fail option
-#' options(old_opts)
+#' # We won't run the remaining examples since they produce equivalent output
+#' if (interactive()) {
```
Comment (schloerke) [inline R/grade_this.R:320]: Should this be a `\dontrun{`?
```suggestion
#' \dontrun{
```

Comment (schloerke) [reply R/grade_this.R:320]: Or update the comment to say the remaining examples produce equivalent output

File: R/grade_this.R [hunk 2]
```diff
@@ -1,84 +1,121 @@
...
+#' # average height of Loblolly pine trees using the `Loblolly` data set.
+#' # We might write an exercise `-check` chunk like the one below.
+#' # 
+#' # Since grade_this() returns a function, we'll save the result of this
+#' # "chunk" as `grader()`, which can be called on an exercise submission
+#' # to evaluate the student's code, which we'll simulate with
+#' # `mock_this_exercise()`.
+#' 
+#' grader <- 
+#'   # ```{r example-check}
```
Comment (schloerke) [inline R/grade_this.R:53]:
```suggestion
#' # ```{r example-check}
```

File: R/graded.R [hunk 3]
```diff
@@ -451,39 +552,94 @@ fail_if <- function(
...
+  user_code = .user_code,
+  solution_code = .solution_code,
   ...,
   env = parent.frame(),
   hint = TRUE,
   encourage = getOption("gradethis.fail.encourage", FALSE),
   allow_partial_matching = getOption("gradethis.allow_partial_matching", TRUE)
 ) {
-  if (is_missing(user_code)) {
+  if (identical(rlang::enexpr(user_code), rlang::expr(.user_code))) {
```
Comment (schloerke) [inline R/graded.R:630]: I understand the motivation here, but I do not like the change.

I don't have a solution that works with both "being able to pass through a missing argument" and "displaying a _comfortable_ default value". 

Let's accept this change for now, but know it comes at the cost of being extendable.

Comment (schloerke) [reply R/graded.R:630]: Same with `solution_code` below

File: R/graded_condition.R [hunk 4]
```diff
@@ -15,11 +15,18 @@ conditionMessage.gradethis_graded <- function(c) {
...
     "[Neutral]"
   }
-
-  paste(
-    correct,
-    paste(as.character(condition_obj$message), collapse = " ")
-  )
+  
+  message <- paste(as.character(condition_obj$message), collapse = " ")
+  len_message <- nchar(message) + nchar(correct) + 3 + nchar("<gradethis_graded>")
```
Comment (schloerke) [inline R/graded_condition.R:20]: Should `len_message` account for internal `\n` in `message`?

**Review by schloerke — APPROVED** [review]

File: R/graded.R (hunk 3 above)
Comment (gadenbuie) [reply R/graded.R:630]: whoops yeah this should use the new placeholder system, I missed this during that refactor. thanks!

File: R/graded_condition.R (hunk 4 above)
Comment (gadenbuie) [reply R/graded_condition.R:20]: Yeah, I'll have those messages wrap regardless if a newline is detected

**Pull request merged**